src/sabre_swap/mod.rs

```diff
@@ -65,72 +67,62 @@ struct TrialResult {
+    front_layer.iter_active().flat_map(move |&v| {
+        neighbors.neighbors[layout.logic_to_phys[v]]
+            .iter()
+            .filter_map(move |&neighbor| {
+                let virtual_neighbor = layout.phys_to_logic[neighbor];
+                if virtual_neighbor > v || !front_layer.is_active(virtual_neighbor) {
+                    // This normalisation line is only necessary to ensure equal output in the
+                    // swap-sorting stage later to the previous version of this algorithm; it can be
+                    // removed when we break that matching.  It isn't needed for determinism.
+                    if v < virtual_neighbor {
+                        Some([v, virtual_neighbor])
+                    } else {
+                        Some([virtual_neighbor, v])
+                    }
                 } else {
-                    [virtual_neighbor, *v]
-                };
-                candidate_swaps.insert(swap);
-            }
-        }
-    }
-    candidate_swaps
+                    None
+                }
+            })
+    })
 }
 
-fn obtain_extended_set(
+/// Fill the given `extended_set` with the next nodes that would be reachable after the front layer
+/// (and themselves).  This uses `required_predecessors` as scratch space for efficiency, but
+/// returns it to the same state as the input on return.
+fn populate_extended_set(
+    extended_set: &mut ExtendedSet,
     dag: &SabreDAG,
-    front_layer: &[NodeIndex],
+    front_layer: &FrontLayer,
     required_predecessors: &mut [u32],
-) -> Vec<[usize; 2]> {
-    let mut extended_set: Vec<[usize; 2]> = Vec::new();
-    let mut decremented: Vec<usize> = Vec::new();
-    let mut tmp_front_layer: Vec<NodeIndex> = front_layer.to_vec();
-    let mut done: bool = false;
-    while !tmp_front_layer.is_empty() && !done {
-        let mut new_tmp_front_layer = Vec::new();
-        for node in tmp_front_layer {
-            for edge in dag.dag.edges(node) {
-                let successor_index = edge.target();
-                let successor = successor_index.index();
-                decremented.push(successor);
-                required_predecessors[successor] -= 1;
-                if required_predecessors[successor] == 0 {
-                    new_tmp_front_layer.push(successor_index);
-                    let node_weight = dag.dag.node_weight(successor_index).unwrap();
-                    let qargs = &node_weight.1;
-                    if qargs.len() == 2 {
-                        let extended_set_edges: [usize; 2] = [qargs[0], qargs[1]];
-                        extended_set.push(extended_set_edges);
-                    }
+) {
+    let mut to_visit = front_layer.iter_nodes().copied().collect::<Vec<_>>();
+    let mut decremented: HashMap<usize, u32> = HashMap::new();
+    let mut i = 0;
+    while i < to_visit.len() && extended_set.len() < EXTENDED_SET_SIZE {
+        for edge in dag.dag.edges_directed(to_visit[i], Direction::Outgoing) {
```

Comment:
fwiw, this would have been fine as `edges()` too because for a directed graph this is the direction behavior in petgraph (which is what the before version did). But being explicit is probably more clear.

Reply:
Ah ok that's good to know thanks.